README.md
libmoost
========

*libmoost* is a collection of C++ utility libraries, including:

* algorithms (set intersection, hashing, variable length encoding, ...)
* abstractions for compiler specific features
* configuration
* container data structures (e.g. LRU cache, memory mapped files)
* message digests
* smart pointers (in addition to boost smart pointers)
* I/O helpers (e.g. async writer, file change watcher)
* key-value store client wrappers for kyoto tycoon, bdb, etc.
* logging
* template metaprogramming
* a stomp message queue client
* object-oriented shared object loading
* service framework (helpers for daemonisation, remote shell access, monitoring, etc.)
* progress bars
* postgres pgq abstraction
* signal handling
* string functions
* unit test support
* threading
* transaction handling
* timers
* xml parser
* complex data structure stringification
* benchmarking
* C++ name demangling
* and more!

Installation
------------

*libmoost* requires a recent C++ compiler (g++, clang++) and boost-1.42 or (ideally) higher. It uses mirbuild as the build system. Once the dependencies are satisfied, you can build and test *libmoost* with

  ./build.py test

and install it using:

  sudo ./build.py install


debian/changelog
lastfm-libmoost (4.0.1) stable; urgency=low

  * Add LICENSE file
  * Add .gitignore
  * Update Doxyfile

 -- Marcus Holland-Moritz <github@mhxnet.de>  Fri, 22 Feb 2013 11:47:20 +0000

lastfm-libmoost (4.0.0) stable; urgency=low

  * Remove monolithic moost library
  * Pull out deprecated http, memcached, postgres code
  * Pull out thrift code that depends on internal projects
  * Refactor moost::which
  * Prepare for public release

 -- Marcus Holland-Moritz <marcus@last.fm>  Tue, 19 Feb 2013 14:46:46 +0000

lastfm-libmoost (3.9.11) stable; urgency=low

  * Fix small issue spotted by clang

 -- Marcus Holland-Moritz <marcus@last.fm>  Tue, 05 Feb 2013 13:01:10 +0000

lastfm-libmoost (3.9.10) stable; urgency=low

  * Make fail_flag_transport work correctly with thrift-0.9.0

 -- Marcus Holland-Moritz <marcus@last.fm>  Wed, 30 Jan 2013 09:18:10 +0000

lastfm-libmoost (3.9.9) stable; urgency=low

  * Add support to moost-mq for reconnection after message
    timeout

 -- Marcus Holland-Moritz <marcus@last.fm>  Mon, 28 Jan 2013 16:31:32 +0000

lastfm-libmoost (3.9.8) stable; urgency=low

  * Correct dependencies of moost packages on lastfm-libmoost-mlog-nsca

 -- Sven Over <sven@last.fm>  Mon, 21 Jan 2013 11:08:53 +0000

lastfm-libmoost (3.9.7) stable; urgency=low

  * Fix unqualified lookup errors with recent gcc versions

 -- Marcus Holland-Moritz <marcus@last.fm>  Fri, 11 Jan 2013 05:44:29 +0000

lastfm-libmoost (3.9.6) stable; urgency=low

  * Remove WIN32 related code from moost::process
  * Fix unqualified lookup errors with recent gcc versions
  * Add libevent-dev as a dependency for libmoost-core-dev

 -- Marcus Holland-Moritz <marcus@last.fm>  Thu, 10 Jan 2013 00:09:25 +0000

lastfm-libmoost (3.9.5) stable; urgency=low

  * Add dependency on fm303 thrift definition, which
    is required to run the thrift compiler to generate
    the test interface definitions

 -- Marcus Holland-Moritz <marcus@last.fm>  Thu, 27 Dec 2012 15:26:53 +0000

lastfm-libmoost (3.9.4) stable; urgency=low

  * Ensure tests run with thrift 0.5.0 and 0.9.0,
    thereby getting rid of check-in generated code
  * Do not call umask(0) in moost::process::daemon_impl
  * Fix integer log2 code for old versions of boost

 -- Marcus Holland-Moritz <marcus@last.fm>  Wed, 26 Dec 2012 21:48:38 +0000

lastfm-libmoost (3.9.3) stable; urgency=low

  * Add moost::math::integer::log2()
  * Remove old murmur hash implementation
  * Add murmur3 hash implementation
  * Add moost::mpl::search()
  * Make operator-> work in safe_shared_ptr
  * Use boost preprocessor library for
    thrift_client_base's execute
  * Small fixes to moost::utils::histogram

 -- Marcus Holland-Moritz <marcus@last.fm>  Wed, 19 Dec 2012 12:26:12 +0000

lastfm-libmoost (3.9.2) stable; urgency=low

  * Add acknowledge support to stomp client library
  * Add default port and logging config support to stomp test client
  * Deprecate dangerous operations on safe_shared_ptr's
  * Improve safe_shared_ptr implementation
  * Add unit tests for async_batch_processor
  * Fix async_batch_processor implementation
  * Move a bunch of threading classes from cyclone to moost

 -- Marcus Holland-Moritz <marcus@last.fm>  Fri, 14 Dec 2012 19:50:30 +0000

lastfm-libmoost (3.9.1) stable; urgency=low

  * Include libmoost_mlog_nsca_appender.so in core package

 -- Marcus Holland-Moritz <marcus@last.fm>  Tue, 04 Dec 2012 19:40:59 +0000

lastfm-libmoost (3.9.0) stable; urgency=low

  * Major cleanup
  * Make moost more modular
  * Got rid of the following components:

      moost/global_scoped_verbose
      moost/utils/is_pod
      moost/igraph
      moost/dynamic_bitset
      moost/boost/asio
      moost/serialization/stdpair_serializer
      moost/algorithm/mersenne_twister
      moost/program_options

  * Integrated pending changes from ricky's branch
  * Fixed issue with root logger being OFF even if enable() was called
  * Integrated thrift logging with the moost logging framework
  * Added mq library with STOMP client
  * Fixed a whole bunch of compiler warnings
  * Added "unused" attribute header
  * Fixed test errors with more recent versions of boost
  * Made moost/shell a wrapper for moost/service/remote_shell
  * Cleaned up include paths and whitespace

 -- Marcus Holland-Moritz <marcus@last.fm>  Tue, 04 Dec 2012 15:12:11 +0000

lastfm-libmoost (3.8.20) stable; urgency=low

  * Moar dependencees

 -- Marcus Holland-Moritz <marcus@last.fm>  Fri, 23 Nov 2012 17:15:47 +0000

lastfm-libmoost (3.8.19) stable; urgency=low

  * Simplify dependency on boost

 -- Marcus Holland-Moritz <marcus@last.fm>  Fri, 23 Nov 2012 15:54:26 +0000

lastfm-libmoost (3.8.18) stable; urgency=low

  * Fix some warnings that show up with -pedantic

 -- Marcus Holland-Moritz <marcus@last.fm>  Fri, 21 Sep 2012 14:52:15 +0000

lastfm-libmoost (3.8.17) stable; urgency=low

  * Make moost::which iterators work with standard containers

 -- Marcus Holland-Moritz <marcus@last.fm>  Thu, 13 Sep 2012 17:44:28 +0000

lastfm-libmoost (3.8.16) stable; urgency=low

  * Build logging nsca appender as shared library
  * Fix debian/control so that libmoost builds on wheezy

 -- Sven Over <sven@last.fm>  Wed, 12 Sep 2012 14:06:36 +0100

lastfm-libmoost (3.8.15) stable; urgency=low

  * Add missing include in kyoto_tycoon_client.hpp

 -- Marcus Holland-Moritz <marcus@last.fm>  Tue, 04 Sep 2012 12:13:22 +0000

lastfm-libmoost (3.8.14) stable; urgency=low

  * Fix potentially undefined behaviour in moost::pdl

 -- Marcus Holland-Moritz <marcus@last.fm>  Wed, 01 Aug 2012 20:50:57 +0000

lastfm-libmoost (3.8.13) stable; urgency=low

  * Add dependency on libgsl and bitmagic

 -- Marcus Holland-Moritz <marcus@last.fm>  Fri, 27 Jul 2012 16:58:44 +0000

lastfm-libmoost (3.8.12) stable; urgency=low

  * Use poll() instead of select() for ionotify implementation
  * Wipe credis from moost

 -- Marcus Holland-Moritz <marcus@last.fm>  Mon, 23 Jul 2012 13:46:15 +0000

lastfm-libmoost (3.8.11) stable; urgency=low


navtree.js
var NAVTREE =
[
  [ "libmoost", "index.html", [
    [ "Related Pages", "pages.html", "pages" ],
    [ "Namespaces", null, [
      [ "Namespace List", "namespaces.html", "namespaces" ],
      [ "Namespace Members", "namespacemembers.html", [
        [ "All", "namespacemembers.html", null ],
        [ "Functions", "namespacemembers_func.html", null ],
        [ "Variables", "namespacemembers_vars.html", null ],
        [ "Typedefs", "namespacemembers_type.html", null ],
        [ "Enumerations", "namespacemembers_enum.html", null ],
        [ "Enumerator", "namespacemembers_eval.html", null ]
      ] ]
    ] ],
    [ "Classes", null, [
      [ "Class List", "annotated.html", "annotated" ],
      [ "Class Index", "classes.html", null ],
      [ "Class Hierarchy", "hierarchy.html", "hierarchy" ],
      [ "Class Members", "functions.html", [
        [ "All", "functions.html", "functions_dup" ],
        [ "Functions", "functions_func.html", "functions_func" ],
        [ "Variables", "functions_vars.html", "functions_vars" ],
        [ "Typedefs", "functions_type.html", "functions_type" ],
        [ "Enumerations", "functions_enum.html", null ],
        [ "Enumerator", "functions_eval.html", null ],
        [ "Related Functions", "functions_rela.html", null ]
      ] ]
    ] ],
    [ "Files", null, [
      [ "File List", "files.html", "files" ],
      [ "File Members", "globals.html", [
        [ "All", "globals.html", "globals_dup" ],
        [ "Functions", "globals_func.html", "globals_func" ],
        [ "Variables", "globals_vars.html", null ],
        [ "Typedefs", "globals_type.html", null ],
        [ "Enumerations", "globals_enum.html", null ],
        [ "Enumerator", "globals_eval.html", null ],
        [ "Defines", "globals_defs.html", "globals_defs" ]
      ] ]
    ] ]
  ] ]
];

function getData(varName)
{
  var i = varName.lastIndexOf('/');
  var n = i>=0 ? varName.substring(i+1) : varName;
  return eval(n);
}

function stripPath(uri)
{
  return uri.substring(uri.lastIndexOf('/')+1);
}

function getScript(scriptName,func,show)
{
  var head = document.getElementsByTagName("head")[0]; 
  var script = document.createElement('script');
  script.id = scriptName;
  script.type = 'text/javascript';
  script.onload = func; 
  script.src = scriptName+'.js'; 
  script.onreadystatechange = function() {
    if (script.readyState == 'complete') { func(); if (show) showRoot(); }
  };
  head.appendChild(script); 
}

function createIndent(o,domNode,node,level)
{
  if (node.parentNode && node.parentNode.parentNode)
  {
    createIndent(o,domNode,node.parentNode,level+1);
  }
  var imgNode = document.createElement("img");
  imgNode.width = 16;
  imgNode.height = 22;
  if (level==0 && node.childrenData)
  {
    node.plus_img = imgNode;
    node.expandToggle = document.createElement("a");
    node.expandToggle.href = "javascript:void(0)";
    node.expandToggle.onclick = function() 
    {
      if (node.expanded) 
      {
        $(node.getChildrenUL()).slideUp("fast");
        if (node.isLast)
        {
          node.plus_img.src = node.relpath+"ftv2plastnode.png";
        }
        else
        {
          node.plus_img.src = node.relpath+"ftv2pnode.png";
        }
        node.expanded = false;
      } 
      else 
      {
        expandNode(o, node, false, false);
      }
    }
    node.expandToggle.appendChild(imgNode);
    domNode.appendChild(node.expandToggle);
  }
  else
  {
    domNode.appendChild(imgNode);
  }
  if (level==0)
  {
    if (node.isLast)
    {
      if (node.childrenData)
      {
        imgNode.src = node.relpath+"ftv2plastnode.png";
      }
      else
      {
        imgNode.src = node.relpath+"ftv2lastnode.png";
        domNode.appendChild(imgNode);
      }
    }
    else
    {
      if (node.childrenData)
      {
        imgNode.src = node.relpath+"ftv2pnode.png";
      }
      else
      {
        imgNode.src = node.relpath+"ftv2node.png";
        domNode.appendChild(imgNode);
      }
    }
  }
  else
  {
    if (node.isLast)
    {
      imgNode.src = node.relpath+"ftv2blank.png";
    }
    else
    {
      imgNode.src = node.relpath+"ftv2vertline.png";
    }
  }
  imgNode.border = "0";
}

function newNode(o, po, text, link, childrenData, lastNode)
{
  var node = new Object();
  node.children = Array();
  node.childrenData = childrenData;
  node.depth = po.depth + 1;
  node.relpath = po.relpath;
  node.isLast = lastNode;

  node.li = document.createElement("li");
  po.getChildrenUL().appendChild(node.li);
  node.parentNode = po;

  node.itemDiv = document.createElement("div");
  node.itemDiv.className = "item";

  node.labelSpan = document.createElement("span");
  node.labelSpan.className = "label";

  createIndent(o,node.itemDiv,node,0);
  node.itemDiv.appendChild(node.labelSpan);
  node.li.appendChild(node.itemDiv);

  var a = document.createElement("a");
  node.labelSpan.appendChild(a);
  node.label = document.createTextNode(text);
  node.expanded = false;
  a.appendChild(node.label);
  if (link) 
  {
    a.className = stripPath(link.replace('#',':'));
    if (link.indexOf('#')!=-1)
    {
      var aname = '#'+link.split('#')[1];
      var srcPage = stripPath($(location).attr('pathname'));
      var targetPage = stripPath(link.split('#')[0]);
      a.href = srcPage!=targetPage ? node.relpath+link : '#';
      a.onclick = function(){
        $('.item').removeClass('selected');
        $('.item').removeAttr('id');
        $(a).parent().parent().addClass('selected');
        $(a).parent().parent().attr('id','selected');
        var anchor = $(aname);
        $("#doc-content").animate({
          scrollTop: anchor.position().top +
          $('#doc-content').scrollTop() -
          $('#doc-content').offset().top
        },500,function(){
          window.location.replace(aname);
        });
      };
    }
    else
    {
      a.href = node.relpath+link;
    }
  } 
  else 
  {
    if (childrenData != null) 
    {
      a.className = "nolink";
      a.href = "javascript:void(0)";
      a.onclick = node.expandToggle.onclick;
    }
  }

  node.childrenUL = null;
  node.getChildrenUL = function() 
  {
    if (!node.childrenUL) 
    {
      node.childrenUL = document.createElement("ul");
      node.childrenUL.className = "children_ul";
      node.childrenUL.style.display = "none";
      node.li.appendChild(node.childrenUL);
    }
    return node.childrenUL;
  };

  return node;
}

function showRoot()
{
  var headerHeight = $("#top").height();
  var footerHeight = $("#nav-path").height();
  var windowHeight = $(window).height() - headerHeight - footerHeight;
  (function (){ // retry until we can scroll to the selected item
    try {
      navtree.scrollTo('#selected',0,{offset:-windowHeight/2});
    } catch (err) {
      setTimeout(arguments.callee, 0);
    }
  })();
}

function expandNode(o, node, imm, showRoot)
{
  if (node.childrenData && !node.expanded) 
  {
    if (typeof(node.childrenData)==='string')
    {
      var varName    = node.childrenData;
      getScript(node.relpath+varName,function(){
        node.childrenData = getData(varName);
        expandNode(o, node, imm, showRoot);
      }, showRoot);
    }
    else
    {
      if (!node.childrenVisited) 
      {
        getNode(o, node);
      }
      if (imm)
      {
        $(node.getChildrenUL()).show();
      } 
      else 
      {
        $(node.getChildrenUL()).slideDown("fast");
      }
      if (node.isLast)
      {
        node.plus_img.src = node.relpath+"ftv2mlastnode.png";
      }
      else
      {
        node.plus_img.src = node.relpath+"ftv2mnode.png";
      }
      node.expanded = true;
    }
  }
}

function showNode(o, node, index)
{
  if (node.childrenData && !node.expanded) 
  {
    if (typeof(node.childrenData)==='string')
    {
      var varName    = node.childrenData;
      getScript(node.relpath+varName,function(){
        node.childrenData = getData(varName);
        showNode(o,node,index);
      },true);
    }
    else
    {
      if (!node.childrenVisited) 
      {
        getNode(o, node);
      }
      $(node.getChildrenUL()).show();
      if (node.isLast)
      {
        node.plus_img.src = node.relpath+"ftv2mlastnode.png";
      }
      else
      {
        node.plus_img.src = node.relpath+"ftv2mnode.png";
      }
      node.expanded = true;
      var n = node.children[o.breadcrumbs[index]];
      if (index+1<o.breadcrumbs.length)
      {
        showNode(o,n,index+1);
      }
      else
      {
        if (typeof(n.childrenData)==='string')
        {
          var varName = n.childrenData;
          getScript(n.relpath+varName,function(){
            n.childrenData = getData(varName);
            node.expanded=false;
            showNode(o,node,index); // retry with child node expanded
          },true);
        }
        else
        {
          if (o.toroot=="index.html" || n.childrenData)
          {
            expandNode(o, n, true, true);
          }
          var a;
          if ($(location).attr('hash'))
          {
            var link=stripPath($(location).attr('pathname'))+':'+
                     $(location).attr('hash').substring(1);
            a=$('.item a[class*=\""'+link+'"\"]');
          }
          if (a && a.length)
          {
            a.parent().parent().addClass('selected');
            a.parent().parent().attr('id','selected');
            var anchor = $($(location).attr('hash'));
            var targetDiv = anchor.next();
            $(targetDiv).children('.memproto,.memdoc').
                     effect("highlight", {}, 1500);
          }
          else
          {
            $(n.itemDiv).addClass('selected');
            $(n.itemDiv).attr('id','selected');
          }
        }
      }
    }
  }
}

function getNode(o, po)
{
  po.childrenVisited = true;
  var l = po.childrenData.length-1;
  for (var i in po.childrenData) 
  {
    var nodeData = po.childrenData[i];
    po.children[i] = newNode(o, po, nodeData[0], nodeData[1], nodeData[2],
      i==l);
  }
}

function initNavTree(toroot,relpath)
{
  var o = new Object();
  o.toroot = toroot;
  o.node = new Object();
  o.node.li = document.getElementById("nav-tree-contents");
  o.node.childrenData = NAVTREE;
  o.node.children = new Array();
  o.node.childrenUL = document.createElement("ul");
  o.node.getChildrenUL = function() { return o.node.childrenUL; };
  o.node.li.appendChild(o.node.childrenUL);
  o.node.depth = 0;
  o.node.relpath = relpath;
  o.node.expanded = false;
  o.node.isLast = true;
  o.node.plus_img = document.createElement("img");
  o.node.plus_img.src = relpath+"ftv2pnode.png";
  o.node.plus_img.width = 16;
  o.node.plus_img.height = 22;

  getScript(relpath+"navtreeindex",function(){
    var navTreeIndex = eval('NAVTREEINDEX');
    if (navTreeIndex) {
      o.breadcrumbs = navTreeIndex[toroot];
      if (o.breadcrumbs==null) o.breadcrumbs = navTreeIndex["index.html"];
      o.breadcrumbs.unshift(0);
      showNode(o, o.node, 0);
    }
  },true);

  $(window).bind('hashchange', function(){
     if (window.location.hash && window.location.hash.length>1){
       var anchor = $(window.location.hash);
       var targetDiv = anchor.next();
       $(targetDiv).children('.memproto,.memdoc').effect("highlight",{},1500);
       var docContent = $('#doc-content');
       if (docContent && anchor && anchor[0] && anchor[0].ownerDocument){
         docContent.scrollTop(anchor.position().top+docContent.scrollTop()-docContent.offset().top);
       }
       var a;
       if ($(location).attr('hash')){
         var link=stripPath($(location).attr('pathname'))+':'+
                  $(location).attr('hash').substring(1);
         a=$('.item a[class*=\""'+link+'"\"]');
       }
       if (a && a.length){
         $('.item').removeClass('selected');
         $('.item').removeAttr('id');
         a.parent().parent().addClass('selected');
         a.parent().parent().attr('id','selected');
         var anchor = $($(location).attr('hash'));
         var targetDiv = anchor.next();
         showRoot();
       }
     } else {
       var docContent = $('#doc-content');
       if (docContent){ docContent.scrollTop(0); }
     }
  })

  $(window).load(showRoot);
}

